GEOSwiftMapboxGL/MapboxGLAnnotations.swift

```diff
@@ -17,112 +24,93 @@ public protocol GEOSwiftMapboxGL {
-The property `shapes` contains MGLShape subclasses instances. When drawing shapes on a map be careful to the fact that that these shapes could be overlays OR annotations.
-*/
-public class MGLShapesCollection : MGLShape, MGLOverlay {
-    let shapes: Array<MGLShape>
-    public let centroid: CLLocationCoordinate2D
-    public let overlayBounds: MGLCoordinateBounds
-    
-    // inserting the where clause in the following generic create some confusion in the precompiler that raise the following error:
-    // Cannot invoke initializer for type ... with an argument list of type (geometryCollection: GeometryCollection<T>)
-    // 1. Expected an argument list of type (geometryCollection: GeometryCollection<T>)
-    required public init<GEOSwiftMapboxGL>(geometryCollection: GeometryCollection<GEOSwiftMapboxGL>) {
-        let shapes = geometryCollection.geometries.map({ (geometry: GEOSwiftMapboxGL) ->
-            MGLShape in
-            return geometry.mapboxShape()
-        })
+extension MGLPolygon {
+  convenience init(linearRing: Polygon.LinearRing) {
+    var coordinates = linearRing.points.map(CLLocationCoordinate2D.init)
+    self.init(coordinates: &coordinates, count: UInt(coordinates.count))
+  }
+  
+  convenience init(polygon: Polygon) {
+    var exteriorCoordinates = polygon.exterior.points.map(CLLocationCoordinate2D.init)
+    self.init(
+      coordinates: &exteriorCoordinates,
+      count: UInt(exteriorCoordinates.count),
+      interiorPolygons: polygon.holes.map(MGLPolygon.init)
+    )
+  }
+}
 
-        if let coordinate = geometryCollection.centroid()?.coordinate {
-            self.centroid = CLLocationCoordinate2DFromCoordinate(coordinate)
-        } else {
-            self.centroid = CLLocationCoordinate2DFromCoordinate(CoordinateFromCLLocationCoordinate2D(CLLocationCoordinate2DMake(0, 0)))
-        }
 
-        self.shapes = shapes
-        
-        if let envelope = geometryCollection.envelope() as? Polygon {
-            let exteriorRing = envelope.exteriorRing
-            let sw = CLLocationCoordinate2DFromCoordinate(exteriorRing.points[0])
-            let ne = CLLocationCoordinate2DFromCoordinate(exteriorRing.points[2])
-            self.overlayBounds = MGLCoordinateBounds(sw:sw, ne:ne)
-        } else {
-            let zeroCoord = CLLocationCoordinate2DMake(0, 0)
-            self.overlayBounds = MGLCoordinateBounds(sw:zeroCoord, ne:zeroCoord)
-        }
 
-        super.init()
-    }
-    
-    required public init?(coder aDecoder: NSCoder) {
-        fatalError("init(coder:) has not been implemented")
-    }
-    
-    override public var coordinate: CLLocationCoordinate2D { get {
-        return centroid
-        }}
-    
-    // TODO: implement using "intersect" method (actually it seems that mapboxgl never calls it...)
-    public func intersects(_ overlayBounds: MGLCoordinateBounds) -> Bool {
-        return true
-    }
+extension Geometry : GEOSwiftMapboxGL {
+    public func mapboxShape() -> MGLShape {
```

Comment:
I think you'd get a more consistent API if you changed this method to be a constructor on `MGLShape` that takes a `GeometryConvertible`. The implementation would be similar. One benefit would be that instead of writing

```
geometries.map { $0.mapboxShape() }
```

you could say

```
geometries.map(MGLShape.init)
```

which feels more consistent with the earlier APIs.

Reply:
Exploring this a little more, it looks like we'll have trouble with init because we can't reassign self. An option that would work, however, is writing it as a factory method in an extension on MGLShape:

```
static func make(with geometry: GeometryConvertible) -> MGLShape
```

Reply:
I think we should also remove the GEOSwiftMapboxGL protocol